core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -538,13 +541,22 @@ class ReplicaManager(val config: KafkaConfig,
           val initialLogEndOffset = localReplica.logEndOffset
           val logReadInfo = localReplica.log match {
             case Some(log) =>
-              val adjustedFetchSize = if (Topic.isInternal(topic) && !readOnlyCommitted) Math.max(fetchSize, log.config.maxMessageSize) else fetchSize
+              val adjustedFetchSize =
+                if (Topic.isInternal(topic) && !readOnlyCommitted)
+                  Math.max(fetchSize, log.config.maxMessageSize)
+                else if (limitBytes > 0)
+                  fetchSize
+                else
+                  0
               log.read(offset, adjustedFetchSize, maxOffsetOpt)
             case None =>
               error("Leader for partition [%s,%d] does not have a local log".format(topic, partition))
               FetchDataInfo(LogOffsetMetadata.UnknownOffsetMetadata, MessageSet.Empty)
           }
 
+          if (limitBytes != Int.MaxValue)
```

Comment:
According to KIP, if limitBytes is equal to Int.MaxValue the fetch request should be processed exactly like pre v.3 request. Maybe we should setup a special flag for this:
val isOldRequest = (limitBytes == Int.MaxValue)
and remove this check